libs/h3/src/main/java/org/elasticsearch/h3/HexRing.java

```diff
@@ -581,52 +599,116 @@ public static long[] hexRing(long origin) {
+                }
+                if (originResDigit >= CoordIJK.Direction.INVALID_DIGIT.digit()) {
+                    // Prevent indexing off the end of the array below
+                    throw new IllegalArgumentException("");
+                }
+                if ((originResDigit == CoordIJK.Direction.K_AXES_DIGIT.digit()
+                    || destinationResDigit == CoordIJK.Direction.K_AXES_DIGIT.digit()) && H3.isPentagon(originParent)) {
+                    // If these are invalid cells, fail rather than incorrectly
+                    // reporting neighbors. For pentagon cells that are actually
+                    // neighbors across the deleted subsequence, they will fail the
+                    // optimized check below, but they will be accepted by the
+                    // gridDisk check below that.
+                    throw new IllegalArgumentException("Undefined error checking for neighbors");
+                }
+                // These sets are the relevant neighbors in the clockwise
+                // and counter-clockwise
+                if (NEIGHBORSETCLOCKWISE[originResDigit].digit() == destinationResDigit
+                    || NEIGHBORSETCOUNTERCLOCKWISE[originResDigit].digit() == destinationResDigit) {
+                    return true;
+                }
+            }
+        }
+        // Otherwise, we have to determine the neighbor relationship the "hard" way.
+        for (int i = 0; i < 6; i++) {
+            long neighbor = h3NeighborRotations(origin, DIRECTIONS[i].digit());
+            if (neighbor != -1) {
+                // -1 is an expected case when trying to traverse off of
+                // pentagons.
+                if (destination == neighbor) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
     /**
      * Returns the hexagon index neighboring the origin, in the direction dir.
      *
-     * Implementation note: The only reachable case where this returns 0 is if the
+     * Implementation note: The only reachable case where this returns -1 is if the
      * origin is a pentagon and the translation is in the k direction. Thus,
-     * 0 can only be returned if origin is a pentagon.
+     * -1 can only be returned if origin is a pentagon.
      *
      * @param origin Origin index
      * @param dir Direction to move in
-     * @param rotations Number of ccw rotations to perform to reorient the
-     *                  translation vector. Will be modified to the new number of
-     *                  rotations to perform (such as when crossing a face edge.)
-     * @param out H3Index of the specified neighbor if succesful
-     * @return E_SUCCESS on success
+     * @return H3Index of the specified neighbor or -1 if there is no more neighbor
      */
-    private static int h3NeighborRotations(long origin, int dir, int[] rotations, long[] out) {
+    private static long h3NeighborRotations(long origin, int dir) {
```

Comment:
Since you removed the previous capability of returning the set of rotations used to find the neighbour, I guess we could also consider renaming the method since it no longer does what its name implies?

Reply:
name changed